doc/developer/design/19042023_stash_migrations.md

```diff
@@ -0,0 +1,352 @@
+exist inside of the `stash/proto/old` folder.
+
+To generate Rust types from our `.proto` schemas, we'd have two steps:
+
+1. Use [`prost_build`](https://docs.rs/prost-build/latest/prost_build/) to do the generation, and
+as part of the build process we'd specifically omit any ["includes"](https://docs.rs/prost-build/latest/prost_build/fn.compile_protos.html)
+which would force our `.proto` files to have no dependencies. 
+2. As part of the build process (i.e. in a `build.rs`) we'd also maintain a hash for the current
+version of the protos (i.e. `stash/proto/objects.proto`) and a hash for each of the snapshotted
+versions. This would allow us to assert at build time that none of our protos have changed. We can
+also assert that we have the correct nubmer of snapshots in `stash/proto/old` based on the 
+`STASH_VERSION` number introduced in the solution to problem 2.
+
+Protobufs also have a [well defined serialization formation](https://protobuf.dev/programming-guides/encoding/)
+so there is no risk of the serialization changing, as long as the `.proto` files themselves don't
+change.
+
+This setup has several benefits:
+
+1. Snapshotting our previous types is very easy. Because we don't specify any "includes" path when
+generating Rust types, the `.proto` files cannot have any dependencies. To snapshot the "current"
+version, all you have to do is copy `objects.proto` and paste it into `/old/objects_vX.proto`.
+2. Detecting at built time if a change has occurred to any of our `.proto` definitions. This goes a
+long way to making the migrations "fearless". If someone changes the current proto definitions, the
+old versions, or bumps the `STASH_VERSION` without snapshotting, we can emit helpful build errors.
+3. It uses "off the shelf" crates, the only code we need to write and thus maintain would be a 
+fairly small `build.rs` script.
+4. **Nice to have**, protobuf has a more compact serialization format than JSON, and is more 
+performant. We also already use protobuf in our codebase, so we're not introducing any new 
+concepts.
+
+
+### Part 2: Structuring Migrations
+
+Along with maintaining `const STASH_VERSION` and using a `match` statement as described above we'd
+setup a pattern of defining migrations in their own modules. Specifically we'd introduce the 
+following layout in our repo:
+
+```
+/stash
+    /src
+        /migrations
+            v0_to_v1.rs
+            v1_to_v2.rs
+            v2_to_v3.rs
+            ...
+        ...
+    /proto
+        ...
+```
+
+and our migration code would then look something like this:
+
+```
+const STASH_VERSION: u64 = 3;
+
+async fn migrate(...) -> Result<...> {
+    let version = stash.read_version().await?;
+
+    const NEXT_VERSION = STASH_VERSION + 1;
+    match version {
+        0 => migrations::v0_to_v1::migrate(...).await?;
+        1 => migrations::v1_to_v2::migrate(...).await?;
+        2 => migrations::v2_to_v3::migrate(...).await?;
```

Comment:
I think this implies that the only type of migrations we support are from versions `X` to `X + 1`, and not arbitrary migrations from `X` to `X + n`. That sounds correct to me, and consistent with our current approach. I just think we should explicitly say that somewhere in this doc.

Reply:
Additionally if that's the case, then we may want an assert somewhere here that `version` is equal to `STASH_VERSION` or `STASH_VERSION - 1`.

Reply:
Good callout, let me add this!